Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
+}
 
-typedef void* gpointer;
+- (void)setUnderlyingQueue:(dispatch_queue_t)queue {
+}
 
-struct NSAtomicListNode {
-    struct NSAtomicListNode* next;
-    StrongId<NSObject> elt;
-};
-typedef struct NSAtomicListNode* NSAtomicListRef;
+- (NSQualityOfService)qualityOfService {
+    return NSQualityOfServiceUserInteractive;
+}
 
-struct NSOperationQueuePriv {
-    NSAtomicListRef myQueues[NSOperationQueuePriority_Count];
-    NSAtomicListRef queues[NSOperationQueuePriority_Count];
+- (void)setQualityOfService:(NSQualityOfService)quality {
+}
+@end
 
-    NSThread* _thread;
-    pthread_mutex_t _threadRunningLock;
+#pragma endregion // NSOperationQueue_MainQueue
 
-    DWORD _maxConcurrentOperationCount;
-    id workAvailable;
-    id suspendedCondition;
-    id allWorkDone;
-    id curOperation;
-    BOOL isSuspended;
+#pragma region NSOperationQueue
 
-    id _name;
+@interface NSOperationQueue ()
+// Private member functions
+- (BOOL)_belowMaxConcurrentOperations;
+- (void)_startOperation:(NSOperation*)operation;
+- (void)_popAndStartQueuedOperations;
+- (void)_setDispatchQueueUsingQualityOfService;
+@end
 
-    NSOperationQueuePriv() {
-        memset(this, 0, sizeof(NSOperationQueuePriv));
-        _threadRunningLock = PTHREAD_MUTEX_INITIALIZER;
-    }
-};
+@implementation NSOperationQueue {
+@private
+    // The actual dispatch queue to which operations are dispatched
+    // Currently, changing between dispatch queues is done by reassigning this member
+    // TODO #:  WinObjC's libdispatch cannot create concurrent dispatch queues (DISPATCH_QUEUE_CONCURRENT just returns a serial queue)
+    //          Because of this, NSOperationQueue has to directly use the global concurrent queues to get concurrent behavior
+    //          Once DISPATCH_QUEUE_CONCURRENT works properly, this should dispatch_set_target instead of reassigning the member
+    dispatch_queue_t _dispatchQueue;
 
-extern pthread_key_t g_currentDispatchQueue;
+    // This backs the public property underlyingQueue. if this is set, _dispatchQueue is set to this
+    dispatch_queue_t _underlyingQueue;
 
-static inline gpointer CompareExchangePointer(volatile gpointer* dest, gpointer exch, gpointer comp) {
-    return (gpointer)EbrCompareExchange((volatile int*)dest, (long)exch, (long)comp);
-}
+    NSQualityOfService _qualityOfService;
 
-NSAtomicListRef NSAtomicListSteal(NSAtomicListRef* listPtr) {
-    NSAtomicListRef ret;
-    do {
-        ret = *listPtr;
-    } while (CompareExchangePointer((void**)listPtr, NULL, ret) != ret);
-    return ret;
-}
+    // Protects _dispatchQueue, _underlyingQueue, and _qualityOfService
+    // Should never be locked before _concurrentOperationCountLock
+    std::recursive_mutex _dispatchQueueLock;
 
-void NSAtomicListReverse(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* cur = *listPtr;
-    struct NSAtomicListNode* prev = NULL;
-    struct NSAtomicListNode* next = NULL;
+    StrongId<NSMutableArray> _operations;
```

Comment:
Not inherently indexed, but the ordering needs to be consistent. Ordered set would be good, I kind of neglected other possibilities just because the operations property is an array. However, we don't actually have NSMutableOrderedSet yet. I'll keep an eye on #2225, and change this to use it once it's in.